odoo/http.py

```diff
@@ -851,581 +1160,429 @@ def render(self, template, qcontext=None, lazy=True, **kw):
-        :rtype: anything
-        """
-        saved_actions = self.get('saved_actions', {})
-        return saved_actions.get("actions", {}).get(key)
-
-    def save_request_data(self):
-        import uuid
-        req = request.httprequest
-        files = werkzeug.datastructures.MultiDict()
-        # NOTE we do not store files in the session itself to avoid loading them in memory.
-        #      By storing them in the session store, we ensure every worker (even ones on other
-        #      servers) can access them. It also allow stale files to be deleted by `session_gc`.
-        for f in req.files.values():
-            storename = 'werkzeug_%s_%s.file' % (self.sid, uuid.uuid4().hex)
-            path = os.path.join(root.session_store.path, storename)
-            with open(path, 'w') as fp:
-                f.save(fp)
-            files.add(f.name, (storename, f.filename, f.content_type))
-        self['serialized_request_data'] = {
-            'form': req.form,
-            'files': files,
-        }
+        return (exc if isinstance(exc, HTTPException)
+           else Forbidden(exc.args[0]) if isinstance(exc, (AccessDenied, AccessError))
+           else BadRequest(exc.args[0]) if isinstance(exc, UserError)
+           else InternalServerError()  # hide the real error
+        )
 
-    @contextlib.contextmanager
-    def load_request_data(self):
-        data = self.pop('serialized_request_data', None)
-        files = werkzeug.datastructures.MultiDict()
-        try:
-            if data:
-                # regenerate files filenames with the current session store
-                for name, (storename, filename, content_type) in data['files'].items():
-                    path = os.path.join(root.session_store.path, storename)
-                    files.add(name, (path, filename, content_type))
-                yield werkzeug.datastructures.CombinedMultiDict([data['form'], files])
-            else:
-                yield None
-        finally:
-            # cleanup files
-            for f, _, _ in files.values():
-                try:
-                    os.unlink(f)
-                except IOError:
-                    pass
 
+class JsonRPCDispatcher(Dispatcher):
+    routing_type = 'json'
 
-def session_gc(session_store):
-    if random.random() < 0.001:
-        # we keep session one week
-        last_week = time.time() - 60*60*24*7
-        for fname in os.listdir(session_store.path):
-            path = os.path.join(session_store.path, fname)
-            try:
-                if os.path.getmtime(path) < last_week:
-                    os.unlink(path)
-            except OSError:
-                pass
+    def __init__(self, request):
+        super().__init__(request)
+        self.jsonrequest = {}
 
-ODOO_DISABLE_SESSION_GC = str2bool(os.environ.get('ODOO_DISABLE_SESSION_GC', '0'))
+    @classmethod
+    def is_compatible_with(cls, request):
+        return request.httprequest.content_type in JSON_MIMETYPES
```

Comment:
```suggestion
        return request.httprequest.mimetype in JSON_MIMETYPES
```
Mime type should be used instead of Content-Type as the later may also include other information like charset (eg. `Content-Type: application/json; charset=utf-8`), like in the mobile apps.

See for reference: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type#syntax